tests/run-tests.py

```diff
@@ -0,0 +1,130 @@
+    osbuild = sp.run(cmd, capture_output=True)
+    if osbuild.returncode != 0:
+        logging.error("osbuild failed!")
+        print("STDERR")
+        print(osbuild.stderr)
+        print("STDOUT")
+        print(osbuild.stdout)
+        if check:
+            sys.exit(1)
+
+    return osbuild.returncode
+
+
+def build_web_server_image():
+    run_osbuild(BASE_INPUT, BASE_OUTPUT, "1-create-base.json")
+    r = run_osbuild(BASE_OUTPUT, WEB_OUTPUT, "2-configure-web-server.json", check=False)
+    # TODO: remove this workaround once the grub2 stage works on the first try :-)
+    if r != 0:
+        run_osbuild(BASE_OUTPUT, WEB_OUTPUT, "2-configure-web-server.json")
+    run_osbuild(WEB_OUTPUT, QCOW2_OUTPUT, "3-compose-qcow2.json")
+
+
+class QemuAcceleration(Enum):
+    KVM = 0
+    HVF = 1
+    NONE = 2
+
+
+class TestResult(Enum):
+    SUCCESS = 0
+    FAIL = 1
+
+
+def uname() -> str:
+    uname = sp.run(["uname"], capture_output=True)
+    return uname.stdout.decode("utf-8").strip()
+
+def detect_qemu_acceleration() -> QemuAcceleration:
+    system = uname()
+    if system == "Linux":
+        lsmod = p = sp.run(["lsmod | grep kvm"], capture_output=True, shell=True)
+        if lsmod.returncode == 0:
+            logging.info("Detected Linux system with KVM support")
+            return QemuAcceleration.KVM
+        else:
+            logging.info("Detected Linux system w/o KVM support")
+            return QemuAcceleration.NONE
+    elif system == "Darwin":
+        logging.info("Detected macOS (support for hvf is expected)")
+        return QemuAcceleration.HVF
+    else:
+        logging.info("Unknown OS")
+        return QemuAcceleration.NONE
+
+
+@contextlib.contextmanager
+def boot_image(path: str):
+    qemu_accel_options = {
+        QemuAcceleration.KVM: ["-accel", "kvm"],
+        QemuAcceleration.HVF: ["-accel", "hvf"],
+        QemuAcceleration.NONE: []
+    }
+    acceleration = detect_qemu_acceleration()
+    if acceleration != QemuAcceleration.NONE:
+        wait_period = EXPECTED_TIME_TO_BOOT_WO_ACCEL
+    else:
+        wait_period = int(EXPECTED_TIME_TO_BOOT_WO_ACCEL / 2)
+    network = ["-net", "nic,model=rtl8139", "-net", "user,hostfwd=tcp::8888-:8888"]
+    cmd = ["qemu-system-x86_64", "-nographic", "-m", "1024", "-snapshot"] + qemu_accel_options[acceleration]\
+          + [path] + network
+    logging.info(f"Booting image: {cmd}")
+    vm = sp.Popen(cmd, stdout=sp.PIPE, stderr=sp.PIPE)
+    try:
+        time.sleep(wait_period)
+        yield None
+    finally:
+        vm.kill()
+
+
+def test_web_server() -> TestResult:
+    cmd = ["curl", "http://127.0.0.1:8888/index"]
+    logging.info(f"Running curl: {cmd}")
+    curl = sp.run(cmd, capture_output=True)
+    logging.info(f"Curl returned: code={curl.returncode}, stdout={curl.stdout}")
+    if curl.returncode == 0 and curl.stdout.decode("utf-8").strip() == "hello, world!":
+        return TestResult.SUCCESS
+    else:
+        return TestResult.FAIL
```

Comment:
I prefer using exceptions for failing tests. That's what python's built-in `unittest` uses as well. Pass `check=True` to the `subprocess.run()` call and you don't need to check the return code at all.

`assert` is very useful in test code and also raises an exception.

Reply:
done